File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        ans = []

        def backtrack(path, target, start):
            if target == 0:
                ans.append(path[:])
                return

            for i in range(start, len(candidates)):
                if i > start and candidates[i] == candidates[i - 1]:
                    continue

                if candidates[i] > target:
                    break

                path.append(candidates[i])
                backtrack(path, target - candidates[i], i + 1)
                path.pop()

        backtrack([], target, 0)
        return ans

File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    void backtrack(vector<vector<int>>&ans,vector<int>&path,vector<int>&candidates,int target,int start)
    {
        if(target==0){
            ans.push_back(path);
            return ;
        }
        for(int i=start;i<candidates.size();i++){
            if(i>start && candidates[i]==candidates[i-1]) continue;
            if(candidates[i]>target) break;
            path.push_back(candidates[i]);
            backtrack(ans,path,candidates,target-candidates[i],i+1);
            path.pop_back();
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>>ans;
        vector<int>path;
        sort(candidates.begin(),candidates.end());
        backtrack(ans,path,candidates,target,0);
        return ans;
    }
};